services/user/Registry/service/src/lib.rs

```diff
@@ -0,0 +1,697 @@
+        let mut metadata = app_metadata_table.get_index_pk().get(&account_id).unwrap();
+        let sender = get_sender();
+        check(
+            account_id == sender || metadata.owners.contains(&sender),
+            "Only app owners can unpublish the app",
+        );
+        check(
+            metadata.status != AppStatus::Published as AppStatusU32,
+            "App is already published",
+        );
+        metadata.status = AppStatus::Published as AppStatusU32;
+        metadata.check_valid();
+        app_metadata_table.put(&metadata).unwrap();
+
+        Wrapper::emit()
+            .history()
+            .appStatusChanged(account_id, metadata.status);
+    }
+
+    #[action]
+    fn unpublish(account_id: AccountNumber) {
+        let app_metadata_table = AppMetadataTable::new();
+        let mut metadata = app_metadata_table.get_index_pk().get(&account_id).unwrap();
+        let sender = get_sender();
+        check(
+            account_id == sender || metadata.owners.contains(&sender),
+            "Only app owners can unpublish the app",
+        );
+        check(
+            metadata.status == AppStatus::Published as AppStatusU32,
+            "App is not published",
+        );
+        metadata.status = AppStatus::Unpublished as AppStatusU32;
+        app_metadata_table.put(&metadata).unwrap();
+
+        Wrapper::emit()
+            .history()
+            .appStatusChanged(account_id, metadata.status);
+    }
+
+    #[action]
+    fn getMetadata(account_id: AccountNumber) -> Option<AppMetadataWithTags> {
+        let app_metadata_table = AppMetadataTable::new();
+        let app_tags_table = AppTagsTable::new();
+        let tags_table = TagsTable::new();
+        let metadata = app_metadata_table.get_index_pk().get(&account_id)?;
+
+        // Get the app tag IDs
+        let tags: Vec<TagRecord> = app_tags_table
+            .get_index_pk()
+            .range((account_id, 0)..(account_id, u32::MAX))
+            .map(|app_tag| {
+                let tag_id = app_tag.tag_id;
+                tags_table.get_index_pk().get(&tag_id).unwrap()
+            })
+            .collect();
+
+        Some(AppMetadataWithTags {
+            metadata: metadata,
+            tags,
+        })
+    }
+
+    #[action]
+    fn getTag(tag_id: u32) -> Option<TagRecord> {
+        let tags_table = TagsTable::new();
+        tags_table.get_index_pk().get(&tag_id)
+    }
+
+    #[action]
+    fn getRelatedTags(partial_tag: String) -> RelatedTags {
+        let table = TagsTable::new();
+        let idx = table.get_index_by_tags();
+
+        let from = partial_tag.to_lowercase();
+        let excluded_to = increment_last_char(from.clone());
+
+        let tags: Vec<String> = idx
+            .range(from..excluded_to)
+            .take(10)
+            .map(|tag_row| tag_row.tag)
+            .collect();
+
+        RelatedTags { tags }
+    }
+
+    #[event(history)]
+    fn appMetadataChanged(app_metadata: AppMetadata) {}
```

Comment:
This is a big event, actually, and not sure how useful it is. I think most similar web apps don't give you a historical view of past descriptions and such. Not to mention you're duplicating data in each event that is impossible to change (`created_at`).

Honestly I'd probably remove this event entirely and just add specific events for specific metadata fields that we expect to want to want to track historically. Which... the only one I can think of is status, and you already have a specific event for that. So maybe just delete this event unless you've got other reasons for it